DOCSP-43419 - GuidRepresentationMode (#235)
mongodb/docs-csharp: source/fundamentals/serialization/guid-serialization.txt

(`GUIDs <https://learn.microsoft.com/en-us/dynamicsax-2012/developer/guids>`__),
also known as **universally unique identifiers** (UUIDs).

> A Short History of MongoDB GUIDs

Reviewer: S: This subheading is a bit too "narrative" maybe something like

> Background on MongoDB GUIDs

or even just

> MongoDB GUIDs

Author: i just want our docs to be more fuuuuuuuuuuun

--------------------------------

A GUID is a 16-byte integer that you can use as a unique ID for a MongoDB document.
The following code block shows an example GUID:

> .. code-block::

Reviewer: S: copyable false

  00112233-4455-6677-8899-aabbccddeeff

Originally, MongoDB represented GUIDs as ``BsonBinaryData``

> values of subtype 3. Because subtype 3 didn't standardize the byte order of GUIDs

Reviewer: S: can link https://www.mongodb.com/docs/manual/reference/bson-types/#binary-data or https://bsonspec.org/spec.html

Author: sure. which info do you think is important to show users?

Author: FWIW, the BsonBinarySubType link on line 72 shows all the subtypes

Reviewer: I think the Server manual pg is better to link

> during encoding, different MongoDB drivers encoded GUIDs with different byte orders.

Reviewer: Q: is there a time frame or C# release # that might be useful to reference here instead of broadly saying "originally"?

Author: i don't know, but i can mark it as a question for the tech reviewer

> Use the following tabs to compare the ways in which different MongoDB language drivers
> encoded the preceding GUID to ``BsonBinaryData`` subtype 3:

Reviewer: ```suggestion
The following tabs contain different driver encodings of the preceding GUID to ``BsonBinaryData`` subtype 3:
```

Author: Don't mind this change, but can you help me understand why it's preferable?

Reviewer: I think it's clearer, less convoluted

.. tabs::

   .. tab:: {+driver-short+}
      :tabid: csharp

>       .. code-block:: csharp

Reviewer: S: make all tabs not copyable

         33221100-5544-7766-8899-aabbccddeeff

   .. tab:: PyMongo
      :tabid: pymongo

      .. code-block:: python

         00112233-4455-6677-8899-aabbccddeeff

   .. tab:: Java Driver
      :tabid: java

      .. code-block:: java

         77665544-3322-1100-ffee-ddccbbaa9988

> To standardize GUID byte order, we added ``BsonBinaryData`` subtype 4, which all
> MongoDB drivers encode in the same way. We recommend using 
> ``BsonBinaryData`` subtype 4 for all new GUIDs.

Reviewer: ```suggestion
We replaced ``BsonBinaryData`` subtype 3 with subtype 4 to standardize the GUID byte order.
We recommend using  ``BsonBinaryData`` subtype 4 to store GUIDs in your application.
```

Author: it's not exactly right to say 4 replaced 3 (since 3 is still an option, albeit one out of favor), so i tried to reword

For a list of all ``BsonBinaryData`` subtypes, see the

> API documentation for the `BsonBinarySubType <{+new-api-root+}/MongoDB.Bson/MongoDB.Bson.BsonBinarySubType.html>`__

Reviewer: Q: checking that this link will be added in v3

Author: yep, it will be broken until the 3.0 docs are published

enum.

> .. tip::

Reviewer: S: add a tip title here

Author: not very descriptive, but i couldn't come up with anything better!

> .. tip::

Reviewer: S: this tip could be in the overview to save users some time if they are wondering which to use

Author: good call

Serializing GUIDs
-----------------

Although we recommend using subtype 4 for all new ``BsonBinaryData`` GUIDs, some older
MongoDB collections might contain some GUID fields that use subtype 3 and others that use
subtype 4. To account for these differences, the {+driver-short+} handles GUID
serialization at the level of individual properties.

If you're using the {+driver-short+} to
:ref:`automap your {+language+} classes to document schemas <csharp-class-mapping>`,
you can add the ``BsonGuidRepresentation`` attribute to a GUID property 
to specify its representation. This attribute accepts a value from the 
``GuidRepresentation`` enum, which contains the following members:

.. list-table::
   :header-rows: 1
   :stub-columns: 1
   :widths: 10 10 

   * - GuidRepresentation Member
     - BsonBinaryData Subtype
   
   * - ``Standard``
     - 4
   
   * - ``CSharpLegacy``
     - 3
   
   * - ``JavaLegacy``
     - 3
   
   * - ``PythonLegacy``
     - 3
   
   * - ``Unspecified``
     - N/A

.. note::
   
   The ``CSharpLegacy``, ``JavaLegacy``, and ``PythonLegacy`` GUID representations are
   all equivalent to ``BsonBinaryData`` subtype 3, but use different byte orders.

> 

Reviewer: S: break out the following content into a sub section

Author: I did a little reorganization, which I think matches your suggestion and improves the overall flow

(`GUIDs <https://learn.microsoft.com/en-us/dynamicsax-2012/developer/guids>`__),
also known as **universally unique identifiers** (UUIDs).

.. tip:: ObjectId

   In MongoDB applications, you can use the
   `ObjectId <{+new-api-root+}/api/MongoDB.Bson/MongoDB.Bson.ObjectId.html>`__ type

>    as a unique identifier for a document. Consider using ``ObjectId`` in place of a GUID

Reviewer: ```suggestion
   as a unique identifier for a document. Consider using ``ObjectId`` instances in place of GUIDs
```

.. tip::

   When you're working with two ``BsonBinaryData`` subtypes, you can combine a global
   serializer with the  ``BsonGuidRepresentation`` property attribute. For example, you
   can register a global serializer for the most commonly used GUID subtype, then use
   the ``BsonGuidRepresentation`` attribute to denote any GUID properties of another subtype.  

.. important::

   If you don't globally register a serializer, you must apply the ``BsonGuidRepresentation``
   attribute to every serializable GUID property. Otherwise, the driver throws an exception
   when it tries to serialize the property.

Serializing Objects
-------------------

You can use an ``ObjectSerializer`` to serialize hierarchical objects to subdocuments. 

> To ensure that GUIDs in these objects are serialized and deserialized correctly,
> you should select the correct GUID representation when constructing your 

Reviewer: s: vale fix
```suggestion
To ensure that GUIDs in these objects are serialized and deserialized correctly,
select the correct GUID representation when constructing your 
```
